flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/autobuffersize/BufferSizeCalculator.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks.autobuffersize;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.TaskManagerOptions;
+import org.apache.flink.runtime.io.network.partition.consumer.IndexedInputGate;
+import org.apache.flink.runtime.mailbox.MailboxExecutor;
+import org.apache.flink.runtime.throughput.ThroughputCalculator;
+import org.apache.flink.streaming.runtime.tasks.TimerService;
+
+import static org.apache.flink.configuration.TaskManagerOptions.AUTOMATIC_BUFFER_ADJUSTMENT_CONSUMPTION_TIME;
+import static org.apache.flink.configuration.TaskManagerOptions.AUTOMATIC_BUFFER_ADJUSTMENT_PERIOD;
+import static org.apache.flink.configuration.TaskManagerOptions.AUTOMATIC_BUFFER_ADJUSTMENT_THRESHOLD_PERCENTAGES;
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Class for automatic calculation of the buffer size based on the current throughput and
+ * configuration.
+ */
+public class BufferSizeCalculator {
+    private static final double MILLIS_IN_SECOND = 1000.0;
+    private final ThroughputCalculator throughputCalculator;
+    private final int bufferAdjustmentPeriod;
+
+    /**
+     * How different should be the total buffer size compare to throughput (when it is 1.0 then
+     * bufferSize == throughput).
+     */
+    private final double throughputFactor;
+
+    private final IndexedInputGate[] inputGates;
+    private final long maxBufferSize;
+    private final long minBufferSize;
+    private final int bufferAdjustmentThresholdPercentages;
+    private final boolean isBufferSizeRecalculationEnabled;
+
+    private long lastBufferSize;
+
+    public BufferSizeCalculator(
+            ThroughputCalculator throughputCalculator,
+            Configuration taskConfig,
+            IndexedInputGate[] inputGates) {
+        this.throughputCalculator = throughputCalculator;
+        this.inputGates = inputGates;
+        this.bufferAdjustmentPeriod = taskConfig.get(AUTOMATIC_BUFFER_ADJUSTMENT_PERIOD);
+        this.throughputFactor =
```

Comment:
rename to `targetBufferSizeInMillis`, `configuredTotalBufferSizeInMillis`?

Reply:
It is not millis it is a factor(or ratio). It shows how many times we should multiply the throughput for having the target buffer size. You can notice that it is `double` not a `long`.